src/sentry/scripts/similarity/index.lua

```diff
@@ -0,0 +1,618 @@
+--[[
+
+Implements an index that can be used to efficiently search for items that
+share similar characteristics.
+
+This implementation is based on MinHash (which is used quickly identify
+similar items and estimate the Jaccard similarity of their characteristic
+sets) but this implementation extends the typical design to add the ability
+to record items by an arbitrary key. This allows querying for similar
+groups that contain many different characteristic sets.
+
+This is modeled as two data structures:
+
+- A bucket frequency hash, which maintains a count of what buckets
+    have been recorded -- and how often -- in a ``(band, key)`` pair. This
+    data can be used to identify what buckets a key is a member of, and also
+    used to identify the degree of bucket similarity when comparing with data
+    associated with another key.
+- A bucket membership set, which maintains a record of what keys have been
+    record in a ``(band, bucket)`` pair. This data can be used to identify
+    what other keys may be similar to the lookup key (but not the degree of
+    similarity.)
+
+]]--
+local function range(start, stop)
+    local result = {}
+    for i = start, stop do
+        table.insert(result, i)
+    end
+    return result
+end
+
+function table.ifilter(t, f)
+    local result = {}
+    for i, value in ipairs(t) do
+        if f(value) then
+            table.insert(result, value)
+        end
+    end
+    return result
+end
+
+function table.imap(t, f)
+    local result = {}
+    for i, value in ipairs(t) do
+        result[i] = f(value)
+    end
+    return result
+end
+
+function table.ireduce(t, f, i)
+    if i == nil then
+        i = {}
+    end
+    local result = i
+    for i, value in ipairs(t) do
+        result = f(result, value)
+    end
+    return result
+end
+
+function table.izip(...)
+    local args = {...}
+    local n = math.max(
+        unpack(
+            table.imap(
+                args,
+                function (t) return #t end
+            )
+        )
+    )
+    local result = {}
+    for i = 1, n do
+        local item = {}
+        for j = 1, #args do
+            table.insert(item, args[j][i])
+        end
+        table.insert(result, item)
+    end
+    return result
+end
+
+function table.slice(t, start, stop)
+    -- NOTE: ``stop`` is inclusive!
+    local result = {}
+    for i = start or 1, stop or #t do
+        table.insert(result, t[i])
+    end
+    return result
+end
+
+
+-- Argument Parsing and Validation
+
+local function parse_number(value)
```

Comment:
I figured it was logically distinct enough that I'd just leave it as independent in case we needed to parse a float, I don't have a strong opinion either way.